rivermigrate/river_migrate.go

```diff
@@ -0,0 +1,502 @@
+//	}
+func New[TTx any](driver riverdriver.Driver[TTx], config *Config) *Migrator[TTx] {
+	if config == nil {
+		config = &Config{}
+	}
+
+	logger := config.Logger
+	if logger == nil {
+		logger = slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+			Level: slog.LevelWarn,
+		}))
+	}
+
+	archetype := &baseservice.Archetype{
+		Logger:     logger,
+		TimeNowUTC: func() time.Time { return time.Now().UTC() },
+	}
+
+	return baseservice.Init(archetype, &Migrator[TTx]{
+		driver:     driver,
+		migrations: riverMigrationsMap,
+		queries:    dbsqlc.New(),
+	})
+}
+
+// MigrateOpts are options for a migrate operation.
+type MigrateOpts struct {
+	// MaxSteps is the maximum number of migrations to apply either up or down.
+	// Leave zero for an unlimited number. Set to -1 to apply no migrations (for
+	// testing/checking purposes).
+	MaxSteps int
+
+	// TargetVersion is a specific migration version to apply migrations to. The
+	// version must exist and it must be in the possible list of migrations to
+	// apply. e.g. If requesting an up migration with version 3, version 3 not
+	// already be applied.
+	//
+	// When applying migrations up, migrations are applied including the target
+	// version, so when starting at version 0 and requesting version 3, versions
+	// 1, 2, and 3 would be applied. When applying migrations down, down
+	// migrations are applied excluding the target version, so when starting at
+	// version 5 an requesting version 3, down migrations for versions 5 and 4
+	// would be applied, leaving the final schema at version 3.
+	TargetVersion int
+}
+
+// MigrateResult is the result of a migrate operation.
+type MigrateResult struct {
+	// Versions are migration versions that were added (for up migrations) or
+	// removed (for down migrations) for this run.
+	Versions []MigrateVersion
+}
+
+// MigrateVersion is the result for a single applied migration.
+type MigrateVersion struct {
+	// Version is the version of the migration applied.
+	Version int
+}
+
+func migrateVersionToInt(version MigrateVersion) int     { return version.Version }
+func migrateVersionToInt64(version MigrateVersion) int64 { return int64(version.Version) }
+
+type Direction string
+
+const (
+	DirectionDown Direction = "down"
+	DirectionUp   Direction = "up"
+)
+
+// Migrate migrates the database in the given direction (up or down). The opts
+// parameter may be omitted for convenience.
+//
+// By default, applies all outstanding migrations possible in either direction.
+// When migrating up all outstanding migrations are applied, and when migrating
+// down all existing migrations are unapplied.
+//
+// When migrating down, use with caution. MigrateOpts.MaxSteps should be set to
+// 1 to only migrate down one step.
+//
+//	res, err := migrator.Migrate(ctx, rivermigrate.DirectionUp, nil)
+//	if err != nil {
+//		// handle error
+//	}
+func (m *Migrator[TTx]) Migrate(ctx context.Context, direction Direction, opts *MigrateOpts) (*MigrateResult, error) {
```

Comment:
Related to the other comment, I'm not quite sure about the API here. I'm thinking in terms of trying to make this foolproof and as difficult as possible to accidentally mess up your prod database, while also being sure exactly what's being run at a given step.

Rather than stating _how many_ migrations to run, do you think it might be safer to specify _which_ migration you want to run? This would require a bit of a different API, because it'd be something like `RunUp(ctx, 1)` or `Run(ctx, rivermigrate.DirectionUp, 3)`. That forces the use case of running multiple migrations into a different API like `RunN()`, but I expect that use case to be more unusual. Also if you went that route, the `N` arg in `RunN()` could be an explicit arg and not something with a default zero value.

My thought process here is that if I'm running an app with River and I see that there's a new version out, I'll check the changelog or website for notes about new migrations in that version. If I see one, I'm going to make a new migration file in my goose/golang-migrate/etc migration setup where I have it call out to `rivermigrate`. I would want to be extra sure that I'm running exactly the migration I expect to be running so that I'm sure I'm prepared and have followed any related instructions if necessary.

Reply:
So both use cases are supported, with what you describe being the `TargetVersion` option. I originally had this even closer to what you described as just `Version` that'd run a single version one off, but I changed it after thinking about the various footguns and ergonomic issues that fell out of being able to target versions that are multiple steps beyond your current version, and would therefore have to error or be applied out of order. `TargetVersion` applies a target version and any versions up to it that are necessary based on your current database state.

I think this would be our general recommendation for usage when writing upgrade guides for specific versions, so the up migration would look like:

``` go
rivermigrate.New(...).Migrate(ctx, rivermigrate.DirectionUp, &riverigrateOpts{
     TargetVersion: 4,
}
```

And then the down migration something like:

``` go
rivermigrate.New(...).Migrate(ctx, rivermigrate.DirectionUp, &riverigrateOpts{
     TargetVersion: 3,
}

// or (neither of these is strictly right when upgrading multiple versions at once — the user would have to make sure to set their original version)

rivermigrate.New(...).Migrate(ctx, rivermigrate.DirectionUp, &riverigrateOpts{
     MaxSteps: 1,
}
```

See also the note above, but I didn't want `TargetVersion` to be the only API because I think there's going to be demand to use this in test suites and one-off small projects (which don't have a full migration line) where you just want to have River raise a usable schema with no hassle and without needing to update code when new River versions come out. Those can just use the really easy:

``` go
rivermigrate.New(...).Migrate(ctx, rivermigrate.DirectionUp, nil)
```

Reply:
Ah yeah I missed this on the first pass, my bad! I think you've highlighted that there are indeed two distinct use cases here: get a working DB as fast as possible, vs give me control and certainty for integrating with my existing migration setup. Iff folks already have an existing DB they likely already have an existing migration setup, which for dev and tests could either consist of stepping through migrations one-at-a-time or doing a bulk schema load.

Since this API supports both, that's great! I want to make sure we at least document these conventions clearly and ensure there aren't any major footguns in them (per above thread):

1. I'm just getting started with River in an existing migration-driven project. What code do I add to a new migration file to go from 0 to latest?
2. I've already adopted river per the above with an older migration version (say, v2). What do I put in a new migration to ensure I go precisely from v2 to v3, no more and no less?
3. I've already adopted river per the above with an older migration version (say, v2) and multiple migrations have come out since my last upgrade. What do I put in a new migration to ensure I go precisely from v2 to i.e. v4, no more and no less?
4. I don't want to integrate my River tables into an existing migration system and just want to use a CLI to quickly bring River to the latest version from wherever it is.
5. I don't want to integrate my River tables into an existing migration system, but I still want the tight control and guarantees from (2) and (3) above.

For (2) and (3) above, would the recommendation be to use _both_ `TargetVersion` and `MaxSteps` together? Or how would I guarantee I'm running only the code from migration 3's up migration, no matter what else I've done?

Reply:
> For (2) and (3) above, would the recommendation be to use _both_ `TargetVersion` and `MaxSteps` together? Or how would I guarantee I'm running only the code from migration 3's up migration, no matter what else I've done?

No, not recommended — by including a `MaxSteps` with everything, you're making things strictly more dangerous.

Say you were trying to get to version 3, and you thought you were on version 2, so you had `MaxSteps: 1`. But actually, you were on version 1! So `TargetVersion: 3, MaxSteps: 1` actually takes you to version 2 and then River might not start up properly.

You might combine these two options while testing something, but you wouldn't want to do so as part of any upgrade guides or the like. `TargetVersion` knows what it's trying to do and will do the right thing based on the current state of the local database.

> For down migrations, I want to be 100% sure that I'm undoing exactly the same steps as I specified above, no more and no less. Does the API let me guarantee that?

Nope, and I don't think there's any way to do so without farming the work out to the user themselves. We can tell people that if they know they're incrementing exactly one version, they can use a `MaxSteps: 1` , but otherwise they'll have to look at their current version and include that as `TargetVersion: <version>`.

One thing I should point out: you generally should never get to the point where you have to run down migrations for any app in production, because if you do, the battle's already lost. River may be an exception here to some degree, but chances are that down migrations are poorly tested and may be unexpectedly destructive/lossy. It's probably less risky to fix forward rather than run them in most cases.

Reply:
> No, not recommended — by including a `MaxSteps` with everything, you're making things strictly more dangerous.
> 
> Say you were trying to get to version 3, and you thought you were on version 2, so you had `MaxSteps: 1`. But actually, you were on version 1! So `TargetVersion: 3, MaxSteps: 1` actually takes you to version 2 and then River might not start up properly.
> 
> You might combine these two options while testing something, but you wouldn't want to do so as part of any upgrade guides or the like. `TargetVersion` knows what it's trying to do and will do the right thing based on the current state of the local database.

Alright, this is a fair point. If you're trying to get to v4 (the current version) and you know this, it should be safe to let the migration engine do what it needs to in order to get to that version. On the down migrations, you could get bit by not migrating down far enough (if you were actually further behind than you thought you were and set a `MaxSteps`) but that could be uncovered through actually testing your up + down migrations together.

The overall design here should be workable as long as we document all these common use cases clearly.